src/ripple/overlay/impl/MessageScheduler.cpp

```diff
@@ -0,0 +1,413 @@
+};
+
+void
+MessageScheduler::connect(std::vector<Channel> channels)
+{
+    if (channels.empty())
+    {
+        return;
+    }
+    std::lock_guard<std::mutex> lock(offersMutex_);
+    if (!senders_.empty())
+    {
+        push_value top(caller, "connect");
+        _offer(channels, senders_);
+    }
+    channels_.reserve(channels_.size() + channels.size());
+    std::move(channels.begin(), channels.end(), std::back_inserter(channels_));
+}
+
+void
+MessageScheduler::disconnect(Peer::id_t peerId)
+{
+    JLOG(journal_.trace()) << "disconnect,id=" << peerId;
+    // We have to acquire both locks for this operation.
+    // Always acquire the offers lock first.
+    push_value top(caller, "disconnect");
+    std::lock_guard<std::mutex> lock(offersMutex_);
+    {
+        auto end = std::remove_if(
+            channels_.begin(), channels_.end(), [peerId](auto const& channel) {
+                return channel.peerId == peerId;
+            });
+        channels_.erase(end, channels_.end());
+    }
+    std::vector<Sender*> senders;
+    {
+        push_value top(tsenders, &senders);
+        std::lock_guard<std::mutex> lock(requestsMutex_);
+        // C++20: Use `std::erase_if`.
+        for (auto it = requests_.begin(); it != requests_.end();)
+        {
+            if (it->second->channel.peerId == peerId)
+            {
+                // This callback may add a sender to `senders`.
+                try
+                {
+                    it->second->receiver->onFailure(
+                        it->second->id, FailureCode::DISCONNECT);
+                }
+                catch (std::exception const&)
+                {
+                }
+                it = requests_.erase(it);
+            }
+            else
+            {
+                ++it;
+            }
+        }
+    }
+    if (!senders.empty())
+    {
+        if (!channels_.empty())
+        {
+            _offer(channels_, senders);
+        }
+        std::move(senders.begin(), senders.end(), std::back_inserter(senders_));
+    }
+}
+
+bool
+MessageScheduler::schedule(Sender* sender)
+{
+    JLOG(journal_.trace()) << "schedule,during=" << caller;
+    if (tsenders)
+    {
+        // The scheduler is already locked in this thread.
+        // Save new senders to be served later.
+        // TODO: Insert in priority order.
+        tsenders->emplace_back(sender);
+        return true;
+    }
+    std::lock_guard<std::mutex> lock(offersMutex_);
+    if (stopped_)
+    {
+        return false;
+    }
+    std::vector<Sender*> senders = {sender};
+    if (!channels_.empty())
+    {
+        push_value top(caller, "schedule");
+        _offer(channels_, senders);
+    }
+    std::move(senders.begin(), senders.end(), std::back_inserter(senders_));
+    return true;
+}
+
+RequestId
+MessageScheduler::send(
+    std::shared_ptr<Peer> peer,
+    protocol::TMGetLedger& message,
+    Receiver* receiver,
+    NetClock::duration timeout)
+{
+    auto requestId = ++nextId_;
```

Comment:
No. It starts at ~16m and wraps after ~4b.

Reply:
Thanks for the explanation @thejohnfreeman.  Probably the most important change would be to add a comment describing the motivation for how `nextId_ ` is handled.  I think it may also be smart to wrap back to 1 << 24 when `nextId_` overflows.  The comment might look something like this:

- The starting value is 1 << 24 or greater to leave enough low numbers so we can reasonably assume that the other senders using the cookie field do not start to creep into our space and unintentionally "reuse" our request identifiers. The other senders use peer identifiers, which are low numbers starting from 1 and incrementing for each peer added, never wrapping, and never reusing even after a peer disconnects.  We never expect the number of peers to come anywhere close to 1 << 24 before a server is shutdown.
- We expect peer requests to happen on the order of 10 per second.  But even if peer requests happen at a rate of 1000/second it would take about 50 days for `nextId_` to overflow.  A timeout should never exceed 24 hours.  So it should be safe to wrap the `nextId_` value back to 1 << 24 if the 32-bit integer ever overflows.